README.md
# HP iLO Metrics Exporter

Blackbox likes exporter used to exports HP Server Integrated Lights Out (iLO) states to Prometheus.

This is a fork from [infinityworks](https://github.com/infinityworks/hpilo-exporter) with some updates/extensions from 
* [dafes](https://github.com/dafes/hpilo-exporter)
* [IDNT](https://github.com/IDNT/hpilo-exporter)


### Gauges

Here are the status code of gauge
```
0 - OK
1 - Degraded
2 - Dead (Other)
```

### Enviroment Variables

The following Enviroment Variables are supported at the moment:

```
ILO_HOST
ILO_PORT
ILO_USER
ILO_PASSWORD
```

### Output example

Example of status of your iLO
```
health_at_a_glance:
  battery: {status: OK}
  bios_hardware: {status: OK}
  fans: {redundancy: Redundant, status: OK}
  memory: {status: OK}
  network: {status: Link Down},
  power_supplies: {redundancy: Redundant, status: OK}
  processor: {status: OK}
  storage: {status: Degraded}
  temperature: {status: OK}
  vrm: {status: Ok}
  drive: {status: Ok}
```

The returned output would be:
```
hpilo_battery{product_name="ProLiant DL360 Gen9",server_name="name.fqdn.domain"} 0.0
hpilo_storage{product_name="ProLiant DL360 Gen9",server_name="name.fqdn.domain"} 1.0
hpilo_fans{product_name="ProLiant DL360 Gen9",server_name="name.fqdn.domain"} 0.0
hpilo_bios_hardware{product_name="ProLiant DL360 Gen9",server_name="name.fqdn.domain"} 0.0
hpilo_memory{product_name="ProLiant DL360 Gen9",server_name="name.fqdn.domain"} 0.0
hpilo_power_supplies{product_name="ProLiant DL360 Gen9",server_name="name.fqdn.domain"} 0.0
hpilo_processor{product_name="ProLiant DL360 Gen9",server_name="name.fqdn.domain"} 0.0
hpilo_network{product_name="ProLiant DL360 Gen9",server_name="name.fqdn.domain"} 2.0
hpilo_temperature{product_name="ProLiant DL360 Gen9",server_name="name.fqdn.domain"} 0.0
hpilo_vrm{product_name="ProLiant DL380 Gen6",server_name="name.fqdn.domain"} 0.0
hpilo_drive{product_name="ProLiant DL380 Gen6",server_name="name.fqdn.domain"} 0.0
hpilo_firmware_version{product_name="ProLiant DL360 Gen9",server_name="name.fqdn.domain"} 2.5
```

#### Temperature

Also the Temperature Sensors are added to the Output

*EXAMPLE*
```
# HELP hpilo_temperature_status HP iLO temperature status
# TYPE hpilo_temperature_status gauge
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="01-Inlet Ambient",server_name="host is unnamed"} 26.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="02-CPU 1",server_name="host is unnamed"} 40.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="03-CPU 2",server_name="host is unnamed"} 40.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="04-P1 DIMM 1-6",server_name="host is unnamed"} 36.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="05-P1 DIMM 7-12",server_name="host is unnamed"} 37.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="06-P2 DIMM 1-6",server_name="host is unnamed"} 38.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="07-P2 DIMM 7-12",server_name="host is unnamed"} 38.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="08-P1 Mem Zone",server_name="host is unnamed"} 34.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="09-P1 Mem Zone",server_name="host is unnamed"} 38.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="10-P2 Mem Zone",server_name="host is unnamed"} 37.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="11-P2 Mem Zone",server_name="host is unnamed"} 34.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="12-HD Max",server_name="host is unnamed"} 35.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="13-Chipset 1",server_name="host is unnamed"} 48.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="14-Chipset1 Zone",server_name="host is unnamed"} 38.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="15-P/S 1 Inlet",server_name="host is unnamed"} 34.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="16-P/S 1 Zone",server_name="host is unnamed"} 36.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="17-P/S 2 Inlet",server_name="host is unnamed"} 36.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="18-P/S 2 Zone",server_name="host is unnamed"} 35.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="21-VR P1",server_name="host is unnamed"} 36.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="22-VR P2",server_name="host is unnamed"} 38.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="23-VR P1 Mem",server_name="host is unnamed"} 30.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="24-VR P1 Mem",server_name="host is unnamed"} 30.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="25-VR P2 Mem",server_name="host is unnamed"} 32.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="26-VR P2 Mem",server_name="host is unnamed"} 31.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="27-VR P1Mem Zone",server_name="host is unnamed"} 29.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="28-VR P1Mem Zone",server_name="host is unnamed"} 29.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="29-VR P2Mem Zone",server_name="host is unnamed"} 31.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="30-VR P2Mem Zone",server_name="host is unnamed"} 31.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="31-HD Controller",server_name="host is unnamed"} 64.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="32-HD Cntlr Zone",server_name="host is unnamed"} 43.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="33-PCI 1 Zone",server_name="host is unnamed"} 38.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="34-PCI 1 Zone",server_name="host is unnamed"} 38.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="36-PCI 2 Zone",server_name="host is unnamed"} 40.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="37-System Board",server_name="host is unnamed"} 40.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="38-System Board",server_name="host is unnamed"} 38.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="39-Sys Exhaust",server_name="host is unnamed"} 37.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="40-Sys Exhaust",server_name="host is unnamed"} 38.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="41-Sys Exhaust",server_name="host is unnamed"} 37.0
hpilo_temperature_status{product_name="ProLiant DL360p Gen8",sensor="42-SuperCAP Max",server_name="host is unnamed"} 29.0
```

### Installing

You can install exporter on the server directly or on separate machine.
To run, you must have `Python3` and `pip3` installed.

To install the requirements with `pip3`:
```
pip3 install -r requirements.txt
```

Then just:
```
python3 ./src/hpilo_exporter/main.py [--address=0.0.0.0 --port=9416 --endpoint="/metrics"]
```



### Docker

Prebuild images are available from the docker repository:
```
dafes/hpilo-exporter:latest
```


To build the image yourself
```
docker build --rm -t hpilo-exporter .
```

To run the container
```
docker run -p 9416:9416 hpilo-exporter:latest
```

You can then call the web server on the defined endpoint, `/metrics` by default.
```
curl 'http://127.0.0.1:9416/metrics?ilo_host=127.0.0.1&ilo_port=443&ilo_user=admin&ilo_password=admin'
```

Passing argument to the docker run command
```
docker run -p 9416:9416 hpilo-exporter:latest --port 9416
```

### Docker compose

Here is an example of Docker Compose deployment:

```yml
version: '3.7'
services:
  hpilo_exporter:
    image: dafes/hpilo-exporter:{{ hpilo_exporter_version }}
    container_name: hpilo_exporter
    ports:
      - 9416:9416
    restart: always
    environment:
      - ILO_USER=FOO
      - ILO_PASSWORD=FOTOPSECRET
```

### Prometheus config

Assuming:
- the exporter is available on `http://localhost:9416`
- you use same the port,username and password for all your iLO

```yml
  - job_name: 'hpilo'
    scrape_interval: 1m
    scrape_timeout: 30s
    params:
      ilo_port: ['443']
      ilo_user: ['FOO']
      ilo_password: ['FOTOPSECRET']
    static_configs:
      - targets: ['192.168.220.188']
        labels:
            service: "ILO"
            host: "ilo_prox1"

    relabel_configs:
      - source_labels: [__address__]
        target_label: __param_ilo_host
      - source_labels: [__param_ilo_host]
        target_label: ilo_host
      - target_label: __address__
        replacement: localhost:9416  # hpilo exporter.
```



CHANGELOG.md
# CHANGELOG
All notable changes to this module will be documented in this file.

## [0.7.1] - 2020-08-03
### Updates
* Loghandlers for Docker Container

## [0.7.0] - 2020-08-03
### Added
* option to specify ilo password and username over enviroment variables

## [0.6.0] - 2020-08-02
### Added
* Function to get temperature values
* ssl context from idnt fork

## [0.5.0] - 2020-08-02
### Updated
* Updated LIB to python3.8
* Updated Dockerfile

## [v0.4.5] - 2018-10-12
### Init
`@ivorcheung ivorcheung tagged this on 12 Oct 2018 · 3 commits to master since this tag`
Version for Pypi

src/hpilo_exporter/main.py
"""
Entrypoint for the application
"""

import argparse
import ssl
import platform

from logger import get_module_logger
from exporter import ILOExporterServer


def main():
    get_module_logger("main").info("HP ILO Metrics Exporter (Python %s, OpenSSL %s)" % (platform.python_version(), ssl.OPENSSL_VERSION))

    parser = argparse.ArgumentParser(description='Exports ilo heath_at_a_glance state to Prometheus')

    parser.add_argument('--address', type=str, dest='address', default='0.0.0.0', help='address to serve on')
    parser.add_argument('--port', type=int, dest='port', default='9416', help='port to bind')
    parser.add_argument('--endpoint', type=str, dest='endpoint', default='/metrics',
                        help='endpoint where metrics will be published')

    args = parser.parse_args()

    exporter = ILOExporterServer(**vars(args))
    exporter.run()


if __name__ == '__main__':
    main()


src/hpilo_exporter/prometheus_metrics.py
from prometheus_client import Gauge
from prometheus_client import REGISTRY

registry = REGISTRY

hpilo_vrm_gauge = Gauge('hpilo_vrm', 'HP iLO vrm status', ["product_name", "server_name"])
hpilo_drive_gauge = Gauge('hpilo_drive', 'HP iLO drive status', ["product_name", "server_name"])
hpilo_battery_gauge = Gauge('hpilo_battery', 'HP iLO battery status', ["product_name", "server_name"])
hpilo_storage_gauge = Gauge('hpilo_storage', 'HP iLO storage status', ["product_name", "server_name"])
hpilo_fans_gauge = Gauge('hpilo_fans', 'HP iLO fans status', ["product_name", "server_name"])
hpilo_bios_hardware_gauge = Gauge('hpilo_bios_hardware', 'HP iLO bios_hardware status', ["product_name", "server_name"])
hpilo_memory_gauge = Gauge('hpilo_memory', 'HP iLO memory status', ["product_name", "server_name"])
hpilo_power_supplies_gauge = Gauge('hpilo_power_supplies', 'HP iLO power_supplies status', ["product_name",
                                                                                            "server_name"])
hpilo_processor_gauge = Gauge('hpilo_processor', 'HP iLO processor status', ["product_name", "server_name"])
hpilo_network_gauge = Gauge('hpilo_network', 'HP iLO network status', ["product_name", "server_name"])
hpilo_temperature_gauge = Gauge('hpilo_temperature', 'HP iLO temperature status', ["product_name", "server_name"])
hpilo_firmware_version = Gauge('hpilo_firmware_version', 'HP iLO firmware version', ["product_name", "server_name"])

hpilo_temperature_status_gauge  = Gauge('hpilo_temperature_status', 'HP iLO temperature status', ["product_name", "server_name", "sensor"])

gauges = {
    'hpilo_vrm_gauge': hpilo_vrm_gauge,
    'hpilo_drive_gauge': hpilo_drive_gauge,
    'hpilo_battery_gauge': hpilo_battery_gauge,
    'hpilo_storage_gauge': hpilo_storage_gauge,
    'hpilo_fans_gauge': hpilo_fans_gauge,
    'hpilo_bios_hardware_gauge': hpilo_bios_hardware_gauge,
    'hpilo_memory_gauge': hpilo_memory_gauge,
    'hpilo_power_supplies_gauge': hpilo_power_supplies_gauge,
    'hpilo_processor_gauge': hpilo_processor_gauge,
    'hpilo_network_gauge': hpilo_network_gauge,
    'hpilo_temperature_gauge': hpilo_temperature_gauge,
    'hpilo_firmware_version': hpilo_firmware_version,
    'hpilo_temperature_status_gauge' : hpilo_temperature_status_gauge
}


src/hpilo_exporter/exporter.py
"""
Pulls data from specified iLO and presents as Prometheus metrics
"""
from __future__ import print_function
from _socket import gaierror
import sys
import hpilo
import ssl
import os

import time
import prometheus_metrics
from http.server import BaseHTTPRequestHandler,HTTPServer
from socketserver import ForkingMixIn
from prometheus_client import generate_latest, Summary
from urllib.parse import parse_qs
from urllib.parse import urlparse

from logger import get_module_logger

def print_err(*args, **kwargs):
    get_module_logger("exporter").error(*args, file=sys.stderr, **kwargs)


# Create a metric to track time spent and requests made.
REQUEST_TIME = Summary(
    'request_processing_seconds', 'Time spent processing request')


class ForkingHTTPServer(ForkingMixIn, HTTPServer):
    max_children = 30
    timeout = 30


class RequestHandler(BaseHTTPRequestHandler):
    """
    Endpoint handler
    """
    def return_error(self):
        self.send_response(500)
        self.end_headers()

    def do_GET(self):
        """
        Process GET request

        :return: Response with Prometheus metrics
        """
        # this will be used to return the total amount of time the request took
        start_time = time.time()
        # get parameters from the URL
        url = urlparse(self.path)
        # following boolean will be passed to True if an error is detected during the argument parsing
        error_detected = False
        query_components = parse_qs(urlparse(self.path).query)

        # 127.0.0.1 - - [03/Aug/2020 10:15:09] "GET /metrics?ilo_host=192.168.220.188&ilo_port=443&ilo_user=prometheus&ilo_password=xzcwjikomEvkqidm4t HTTP/1.1" 200 -

        get_module_logger("exporter").info("{} GET {}".format(self.client_address ,self.path))

        ilo_host = None
        ilo_port = None
        ilo_user = None
        ilo_password = None
        try:
            ilo_host = query_components.get('ilo_host', [''])[0] or os.environ['ILO_HOST']
            ilo_port = int(query_components.get('ilo_port', [''])[0] or os.environ['ILO_PORT'])
            ilo_user = query_components.get('ilo_user', [''])[0] or os.environ['ILO_USER']
            ilo_password = query_components.get('ilo_password', [''])[0] or os.environ['ILO_PASSWORD']

        except KeyError as e:
            get_module_logger("exporter").error("missing parameter %s" % e)
            self.return_error()
            error_detected = True

        if url.path == self.server.endpoint and ilo_host and ilo_user and ilo_password and ilo_port:

            ilo = None
            ssl_context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
            # Sadly, ancient iLO's aren't dead yet, so let's enable sslv3 by default
            ssl_context.options &= ~ssl.OP_NO_SSLv3
            ssl_context.check_hostname = False
            ssl_context.set_ciphers(('ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:ECDH+HIGH:'
            'DH+HIGH:ECDH+3DES:DH+3DES:RSA+AESGCM:RSA+AES:RSA+HIGH:RSA+3DES:!aNULL:'
            '!eNULL:!MD5'))

            try:
                ilo = hpilo.Ilo(hostname=ilo_host,
                                login=ilo_user,
                                password=ilo_password,
                                port=ilo_port, timeout=10,
                                ssl_context=ssl_context)
            except hpilo.IloLoginFailed:
                get_module_logger("exporter").error("ILO login failed")
                self.return_error()
            except gaierror:
                get_module_logger("exporter").error("ILO invalid address or port")
                self.return_error()
            except hpilo.IloCommunicationError as e:
                get_module_logger("exporter").error(e)
                self.return_error()

            # get product and server name
            try:
                product_name = ilo.get_product_name()
            except:
                product_name = "Unknown HP Server"

            try:
                server_name = ilo.get_server_name()
                if server_name == "":
                    server_name = ilo_host
            except:
                server_name = ilo_host

            # get health
            embedded_health = ilo.get_embedded_health()
            health_at_glance = embedded_health['health_at_a_glance']
            


            if health_at_glance is not None:
                for key, value in health_at_glance.items():
                    for status in value.items():
                        if status[0] == 'status':
                            gauge = 'hpilo_{}_gauge'.format(key)
                            if status[1].upper() == 'OK':
                                prometheus_metrics.gauges[gauge].labels(product_name=product_name,
                                                                        server_name=server_name).set(0)
                            elif status[1].upper() == 'DEGRADED':
                                prometheus_metrics.gauges[gauge].labels(product_name=product_name,
                                                                        server_name=server_name).set(1)
                            else:
                                prometheus_metrics.gauges[gauge].labels(product_name=product_name,
                                                                        server_name=server_name).set(2)

            # get firmware version
            fw_version = ilo.get_fw_version()["firmware_version"]
            # prometheus_metrics.hpilo_firmware_version.set(fw_version)
            prometheus_metrics.hpilo_firmware_version.labels(product_name=product_name,
                                                             server_name=server_name).set(fw_version)
            # get temperature informations
            for temp in embedded_health['temperature']:
                value = embedded_health['temperature'][temp]['currentreading'][0]
                if value != "N":
                    prometheus_metrics.hpilo_temperature_status_gauge.labels(product_name=product_name,
                                                            server_name=server_name,
                                                            sensor=temp).set(value)


            # get the amount of time the request took
            REQUEST_TIME.observe(time.time() - start_time)
            get_module_logger("exporter").info("REQUEST_TIME: {}s".format(str(REQUEST_TIME._sum._value)))

            # generate and publish metrics
            metrics = generate_latest(prometheus_metrics.registry)
            self.send_response(200)
            self.send_header('Content-Type', 'text/plain')
            self.end_headers()
            self.wfile.write(metrics)

        elif url.path == '/':
            self.send_response(200)
            self.send_header('Content-Type', 'text/html')
            self.end_headers()
            self.wfile.write("""<html>
            <head><title>HP iLO Exporter</title></head>
            <body>
            <h1>HP iLO Exporter</h1>
            <p>Visit <a href="/metrics">Metrics</a> to use.</p>
            </body>
            </html>""")

        else:
            if not error_detected:
                self.send_response(404)
                self.end_headers()

    def log_message(self, format, *args):
        return

class ILOExporterServer(object):
    """
    Basic server implementation that exposes metrics to Prometheus
    """

    def __init__(self, address='0.0.0.0', port=8080, endpoint="/metrics"):
        self._address = address
        self._port = port
        self.endpoint = endpoint

    def print_info(self):
        get_module_logger("exporter").info("Starting exporter on: http://{}:{}{}".format(self._address, self._port, self.endpoint))

    def run(self):
        self.print_info()

        server = ForkingHTTPServer((self._address, self._port), RequestHandler)
        server.endpoint = self.endpoint

        try:
            while True:
                server.handle_request()
        except KeyboardInterrupt:
            get_module_logger("exporter").info("Killing exporter")
            server.server_close()
